tests/components/tasmota/test_common.py

```diff
@@ -0,0 +1,328 @@
+    config1[CONF_PREFIX][PREFIX_TELE] = "tele1"
+    config1[CONF_OFFLINE] = "offline1"
+    config1[CONF_ONLINE] = "online1"
+    config2 = copy.deepcopy(config)
+    config2[CONF_PREFIX][PREFIX_TELE] = "tele2"
+    config2[CONF_OFFLINE] = "offline2"
+    config2[CONF_ONLINE] = "online2"
+    data1 = json.dumps(config1)
+    data2 = json.dumps(config2)
+
+    availability_topic1 = get_topic_tele_will(config1)
+    availability_topic2 = get_topic_tele_will(config2)
+    assert availability_topic1 != availability_topic2
+    offline1 = get_state_offline(config1)
+    offline2 = get_state_offline(config2)
+    assert offline1 != offline2
+    online1 = get_state_online(config1)
+    online2 = get_state_online(config2)
+    assert online1 != online2
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config1[CONF_MAC]}/config", data1)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    async_fire_mqtt_message(hass, availability_topic1, online1)
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+    async_fire_mqtt_message(hass, availability_topic1, offline1)
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    # Change availability settings
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config2[CONF_MAC]}/config", data2)
+    await hass.async_block_till_done()
+
+    # Verify we are no longer subscribing to the old topic or payload
+    async_fire_mqtt_message(hass, availability_topic1, online1)
+    async_fire_mqtt_message(hass, availability_topic1, online2)
+    async_fire_mqtt_message(hass, availability_topic2, online1)
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    # Verify we are subscribing to the new topic
+    async_fire_mqtt_message(hass, availability_topic2, online2)
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+
+async def help_test_discovery_removal(
+    hass, mqtt_mock, caplog, domain, config1, config2
+):
+    """Test removal of discovered entity."""
+    device_reg = await hass.helpers.device_registry.async_get_registry()
+    entity_reg = await hass.helpers.entity_registry.async_get_registry()
+
+    data1 = json.dumps(config1)
+    data2 = json.dumps(config2)
+    assert config1[CONF_MAC] == config2[CONF_MAC]
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config1[CONF_MAC]}/config", data1)
+    await hass.async_block_till_done()
+
+    # Verify device and entity registry entries are created
+    device_entry = device_reg.async_get_device(set(), {("mac", config1[CONF_MAC])})
+    assert device_entry is not None
+    entity_entry = entity_reg.async_get(f"{domain}.test")
+    assert entity_entry is not None
+
+    # Verify state is added
+    state = hass.states.get(f"{domain}.test")
+    assert state is not None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)